src/System.Net.WebSockets.WebSocketProtocol/src/System/Net/WebSockets/ManagedWebSocketExtensions.cs

```diff
@@ -0,0 +1,62 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.IO;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.WebSockets
+{
+    internal static class ManagedWebSocketExtensions
+    {
+        internal static unsafe string GetString(this UTF8Encoding encoding, Span<byte> bytes)
+        {
+            fixed (byte* b = &MemoryMarshal.GetReference(bytes))
+            {
+                return encoding.GetString(b, bytes.Length);
+            }
+        }
+
+        internal static Task<int> ReadAsync(this Stream stream, Memory<byte> destination, CancellationToken cancellationToken)
+        {
+            if (destination.TryGetArray(out ArraySegment<byte> array))
+            {
+                return stream.ReadAsync(array.Array, array.Offset, array.Count, cancellationToken);
+            }
+            else
+            {
+                byte[] buffer = ArrayPool<byte>.Shared.Rent(destination.Length);
+                return FinishReadAsync(stream.ReadAsync(buffer, 0, destination.Length, cancellationToken), buffer, destination);
+
+                async Task<int> FinishReadAsync(Task<int> readTask, byte[] localBuffer, Memory<byte> localDestination)
+                {
+                    try
+                    {
+                        int result = await readTask.ConfigureAwait(false);
+                        new Span<byte>(localBuffer, 0, result).CopyTo(localDestination.Span);
+                        return result;
+                    }
+                    finally
+                    {
+                        ArrayPool<byte>.Shared.Return(localBuffer);
+                    }
+                }
+            }
+        }
+    }
+
+    internal static class BitConverter
+    {
+        internal static unsafe int ToInt32(ReadOnlySpan<byte> value)
+        {
+            if (value.Length < sizeof(int))
+                throw new ArgumentOutOfRangeException(nameof(value));
```

Comment:
Based on the call sites, will this ever fail? Wondering if it can instead be an assert.

Reply:
The function that calls to the BitConverter takes a byte array, and slices it based on an offset. It could be the case that length-offset < sizeof(int).. The byte array is _sendBuffer which is atleast 14 bytes in length, and markOffset comes from WriteHeader.

Reply:
> It could be the case that length-offset < sizeof(int)

When? You're saying there are reachable cases today where this results exceptions being thrown?

Reply:
I am not sure, it depends upon what the runtime values of markOffset and _sendBuffer could be..

Reply:
I'm suggesting we validate that indeed the inputs will always be valid and then change this to an assert.  If this did throw an exception, it indicates a problem in earlier logic that should have guaranteed there was enough data for this to be used.

Reply:
There is a debug.assert in applymask that checks the conditions of the bytearray and maskoffset before calling the method that calls this code, so this can be changed to assert.